Sources/NIO/EventLoop.swift

```diff
@@ -580,21 +588,45 @@ extension EventLoop {
         return EventLoopIterator([self])
     }
 
-    /// Checks that this call is run from the EventLoop. If this is called from within the EventLoop this function will
-    /// have no effect, if called from outside the EventLoop it will crash the process with a trap if run in debug mode.
-    /// In release mode this function never has any effect.
+    /// Asserts that the current thread is the one tied to this `EventLoop`.
+    /// Otherwise, if running in debug mode, the process will be abnormally terminated as per the semantics of
+    /// `preconditionFailure(_:file:line:)`. Never has any effect in release mode.
     ///
     /// - note: This is not a customization point so calls to this function can be fully optimized out in release mode.
     @inlinable
-    public func assertInEventLoop(file: StaticString = #file, line: UInt = #line) {
+    public func assertInEventLoop(_ message: @autoclosure() -> String = "", file: StaticString = #file, line: UInt = #line) {
         debugOnly {
-            self.preconditionInEventLoop(file: file, line: line)
+            self.preconditionInEventLoop(message(), file: file, line: line)
+        }
+    }
+
+    /// Asserts that the current thread is _not_ the one tied to this `EventLoop`.
+    /// Otherwise, if running in debug mode, the process will be abnormally terminated as per the semantics of
+    /// `preconditionFailure(_:file:line:)`. Never has any effect in release mode.
+    ///
+    /// - note: This is not a customization point so calls to this function can be fully optimized out in release mode.
+    @inlinable
+    public func assertNotInEventLoop(_ message: @autoclosure() -> String = "", file: StaticString = #file, line: UInt = #line) {
+        debugOnly {
+            self.preconditionNotInEventLoop(message(), file: file, line: line)
         }
     }
 
     /// Checks the necessary condition of currently running on the called `EventLoop` for making forward progress.
+    /// Forwarder to preserve API while adding message parameter.
+    @inlinable
     public func preconditionInEventLoop(file: StaticString = #file, line: UInt = #line) {
-        precondition(self.inEventLoop, file: file, line: line)
+        preconditionInEventLoop("", file: file, line: line)
+    }
+
+    /// Checks the necessary condition of currently running on the called `EventLoop` for making forward progress.
+    public func preconditionInEventLoop(_ message: @autoclosure() -> String, file: StaticString = #file, line: UInt = #line) {
+        precondition(self.inEventLoop, message(), file: file, line: line)
+    }
+
+    /// Checks the necessary condition of currently _not_ running on the called `EventLoop` for making forward progress.
+    public func preconditionNotInEventLoop(_ message: @autoclosure() -> String = "", file: StaticString = #file, line: UInt = #line) {
```

Comment:
This should be inlinable, or else we get needless allocations in Swift 5.0 for example.